src/Microsoft.DocAsCode.Common/CompositeDictionary.cs

```diff
@@ -0,0 +1,245 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DocAsCode.Common
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.Collections.Immutable;
+    using System.Linq;
+
+    public class CompositeDictionary
+        : IDictionary<string, object>
+    {
+
+        private readonly ImmutableArray<Entry> _entries;
+
+        private CompositeDictionary(ImmutableArray<Entry> entries)
+        {
+            _entries = entries;
+        }
+
+        public object this[string key]
+        {
+            get
+            {
+                if (key == null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+                foreach (var entry in _entries)
+                {
+                    if (key.StartsWith(entry.Prefix))
+                    {
+                        return entry.GetValue(key.Substring(entry.Prefix.Length));
+                    }
+                }
+                throw new KeyNotFoundException();
+            }
+            set
+            {
+                if (key == null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+                foreach (var entry in _entries)
+                {
+                    if (key.StartsWith(entry.Prefix))
+                    {
+                        entry.SetValue(key.Substring(entry.Prefix.Length), value);
+                        return;
+                    }
+                }
+            }
```

Comment:
I mean whether should set the value when the key not starts with prefix?

---
In reply to: [97214579](https://github.com/dotnet/docfx/pull/1216#discussion_r97214579) [](ancestors = 97214579)